pkg/controller/gcp/compute/globaladdress.go

```diff
@@ -0,0 +1,148 @@
+	errGetAddress       = "cannot get external Address resource"
+	errCreateAddress    = "cannot create external Address resource"
+	errDeleteAddress    = "cannot delete external Address resource"
+	errUpdateManaged    = "cannot update managed resource"
+)
+
+// GlobalAddressController is the controller for GlobalAddress CRD.
+type GlobalAddressController struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *GlobalAddressController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(strings.ToLower(fmt.Sprintf("%s.%s", v1alpha1.GlobalAddressKindAPIVersion, v1alpha1.Group))).
+		For(&v1alpha1.GlobalAddress{}).
+		Complete(resource.NewManagedReconciler(mgr,
+			resource.ManagedKind(v1alpha1.GlobalAddressGroupVersionKind),
+			resource.WithExternalConnecter(&gaConnector{client: mgr.GetClient(), newCompute: compute.NewService}),
+			resource.WithManagedConnectionPublishers()))
+}
+
+type gaConnector struct {
+	client     client.Client
+	newCompute func(ctx context.Context, opts ...option.ClientOption) (*compute.Service, error)
+}
+
+func (c *gaConnector) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	ga, ok := mg.(*v1alpha1.GlobalAddress)
+	if !ok {
+		return nil, errors.New(errNotGlobalAddress)
+	}
+
+	p := &gcpv1alpha1.Provider{}
+	if err := c.client.Get(ctx, meta.NamespacedNameOf(ga.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errProviderNotRetrieved)
+	}
+	s := &v1.Secret{}
+	if err := c.client.Get(ctx, types.NamespacedName{Namespace: p.Namespace, Name: p.Spec.Secret.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errProviderSecretNotRetrieved)
+	}
+
+	cmp, err := c.newCompute(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(compute.ComputeScope))
+	return &gaExternal{client: c.client, compute: cmp, projectID: p.Spec.ProjectID}, errors.Wrap(err, errNewClient)
+}
+
+type gaExternal struct {
+	client    client.Client
+	compute   *compute.Service
+	projectID string
+}
+
+func (e *gaExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) {
+	ga, ok := mg.(*v1alpha1.GlobalAddress)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotGlobalAddress)
+	}
+	observed, err := e.compute.GlobalAddresses.Get(e.projectID, ga.Spec.Name).Context(ctx).Do()
+	if gcp.IsErrorNotFound(err) {
+		return resource.ExternalObservation{ResourceExists: false}, nil
+	}
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(err, errGetAddress)
+	}
+
+	globaladdress.UpdateStatus(&ga.Status, observed)
+	globaladdress.UpdateParameters(&ga.Spec.GlobalAddressParameters, observed)
+
+	// Global addresses are always "up to date" because they can't be updated. ¯\_(ツ)_/¯
```

Comment:
I'm thinking of the following scenario: I updated either `Subnetwork` or some other field. The controller never updated the resource on the external provider, because it's not possible. But at the same time nothing stopped me from doing that. That is OK unless if I want to go back and extract actual value of `Subnetwork` at some point in the future. That information, on our level, is lost. The spec is not propagated and state doesn't have that field. So, we either;
- Need to include _all_ fields in the status to provide a precise and guaranteed state of the resource in the provider
- Guarantee that spec will always be true and if some value cannot be updated we don't allow it. If it can be updated, then state is the place you should be looking at.

IMO, second option is harder to maintain. Crossplane needs to make sure that the fields that are immutable are immutable, the client of the resource should know which information should be extracted from which part (spec or state). Note that not all clients of the CRs are in golang, so they may not have the type and field information of fields in spec and state (thinking of bash scripts).

In the first scenario, the statement is that whatever you see in the state is the current status that we can know. In the future, we can even limit RBAC rules that consumers of the CRs to allow them to see only subresource `status` and that would be enough for them. https://kubernetes.io/docs/reference/access-authn-authz/rbac/#command-line-utilities

Reply:
I agree that we currently have a bad user experience here. This is not by design, but rather because it is one of the many areas we haven't yet had time to address. We're still in alpha for a reason. :)

We can work through this in more detail as part of your impending custom resource one pager, but I still feel quite strongly that including all fields in the status is not the right path. I think we should mark fields that cannot be changed as immutable. We're tracking this in https://github.com/crossplaneio/crossplane/issues/727 - there is a KEP open to add support for flagging fields as immutable at the CRD level.